Userland/Applications/Presenter/SlideObject.h

```diff
@@ -0,0 +1,137 @@
+
+    virtual void paint(Gfx::Painter&, Gfx::FloatSize display_scale) const;
+    ALWAYS_INLINE Gfx::IntRect transformed_bounding_box(Gfx::IntRect clip_rect, Gfx::FloatSize display_scale) const;
+
+    void set_rect(Gfx::IntRect rect) { m_rect = rect; }
+    Gfx::IntRect rect() const { return m_rect; }
+
+protected:
+    SlideObject();
+
+    Gfx::IntRect m_rect;
+};
+
+// Objects with a foreground color.
+class GraphicsObject : public SlideObject {
+    C_OBJECT_ABSTRACT(SlideObject);
+
+public:
+    virtual ~GraphicsObject() = default;
+
+    void set_color(Gfx::Color color) { m_color = color; }
+    Gfx::Color color() const { return m_color; }
+
+protected:
+    GraphicsObject();
+
+    Gfx::Color m_color;
+};
+
+class Text : public GraphicsObject {
+    C_OBJECT(SlideObject);
+
+public:
+    Text();
+    virtual ~Text() = default;
+
+    virtual void paint(Gfx::Painter&, Gfx::FloatSize display_scale) const override;
+
+    void set_font(String font) { m_font = move(font); }
+    StringView font() const { return m_font; }
+    void set_font_size(int font_size) { m_font_size = font_size; }
+    int font_size() const { return m_font_size; }
+    void set_font_weight(unsigned font_weight) { m_font_weight = font_weight; }
+    unsigned font_weight() const { return m_font_weight; }
+    void set_text_alignment(Gfx::TextAlignment text_alignment) { m_text_alignment = text_alignment; }
+    Gfx::TextAlignment text_alignment() const { return m_text_alignment; }
+    void set_text(String text) { m_text = move(text); }
+    StringView text() const { return m_text; }
+
+protected:
+    String m_text;
+    // The font family, technically speaking.
+    String m_font;
+    int m_font_size;
+    unsigned m_font_weight;
+    Gfx::TextAlignment m_text_alignment;
+};
+
+// How to scale an image object.
+enum class ImageScaling {
+    // Fit the image into the bounding box, preserving its aspect ratio.
+    FitSmallest,
+    // Match the bounding box in width and height exactly; this will change the image's aspect ratio if the aspect ratio of the bounding box is not exactly the same.
+    Stretch,
+    // Make the image fill the bounding box, preserving its aspect ratio. This means that the image will be cut off on the top and bottom or left and right, depending on which dimension is "too large".
+    FitLargest,
+};
+
+class Image : public SlideObject {
+    C_OBJECT(Image);
+
+public:
+    Image(NonnullRefPtr<ImageDecoderClient::Client>, NonnullRefPtr<GUI::Window>);
+    virtual ~Image() = default;
+
+    virtual void paint(Gfx::Painter&, Gfx::FloatSize display_scale) const override;
+
+    void set_image_path(String image_path)
+    {
+        m_image_path = move(image_path);
+        auto result = reload_image();
+        if (result.is_error())
+            GUI::MessageBox::show_error(m_window, String::formatted("Loading image {} failed: {}", m_image_path, result.error()));
```

Comment:
if these objects were widgets, you could just call window() instead of having to plumb  a GUI window through the hierarchy. with the only use being showing this message box.

Reply:
Again, I'll change how images are loaded in the future so this code is just to get things going.